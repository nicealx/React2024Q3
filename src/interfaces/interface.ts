export interface IPeople {
  name: string;
  birth_year: string;
  eye_color: string;
  gender: string;
  hair_color: string;
  height: string;
  mass: string;
  skin_color: string;
  homeworld: string;
  films: string[];
  species: string[];
  starships: string[];
  vehicles: string[];
  url: string;
  created: string;
  edited: string;
}

export interface IData {
  count: number;
  next: string | null;
  previous: string | null;
  results: IPeople[];
}

export type IDataCard = Pick<
  IPeople,
  'name' | 'hair_color' | 'height' | 'mass'
>;

export interface IAppContext {
  isLoading: boolean;
  cards: IPeople[];
  searchText: string;
  search: (value: string) => Promise<void>;
}
