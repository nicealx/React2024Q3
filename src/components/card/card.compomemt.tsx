import { Component } from 'react';
import { IDataCard } from '../../interfaces/interface';
import './card.css';

interface ICard {
  data: IDataCard;
}

export class Card extends Component<ICard> {
  render() {
    return (
      <div className="card__item">
        <div className="card__content">
          <h2 className="card__title">{this.props.data.name}</h2>
          <p>
            <span className="card__span">Heigth:</span> {this.props.data.height}
          </p>
          <p>
            <span className="card__span">Mass:</span> {this.props.data.mass}
          </p>
          <p>
            <span className="card__span">Hair color:</span>{' '}
            {this.props.data.hair_color}
          </p>
        </div>
        <div className="card__backdrop">{this.props.data.name}</div>
      </div>
    );
  }
}
