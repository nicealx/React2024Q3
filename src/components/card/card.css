.card__item {
  color: inherit;
}

.card__content {
  z-index: 2;
}

.card__title {
  margin: 10px 0;
  transition: color 0.5s ease;
}

.card__backdrop {
  position: absolute;
  top: 50%;
  left: 50%;
  font-weight: 700;
  transform: translate(-50%, -50%);
  opacity: 0;
  transition:
    opacity 0.5s ease,
    transform 0.5s ease;
  z-index: 1;
}

.card__item:hover .card__title,
.card__item.active {
  color: blue;
}

.card__item:hover .card__backdrop,
.card__item.active .card__backdrop {
  opacity: 0.2;
  transform: translate(-50%, -50%) scale(5) rotate(-20deg);
}
