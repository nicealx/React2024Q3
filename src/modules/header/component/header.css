.header {
  display: flex;
  flex-direction: column;
  gap: 20px;
  width: 100%;
  align-self: flex-start;
}
