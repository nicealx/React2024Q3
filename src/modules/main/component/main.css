.main {
  display: flex;
  gap: 20px;
  flex-grow: 1;
}

.main__wrap {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  gap: 20px;
  flex-grow: 1;
}
