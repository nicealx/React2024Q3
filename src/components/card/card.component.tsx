import { FC } from 'react';
import { IPeople } from '../../interfaces/interface';
import { BASE_URL } from '../../config/constants';
import { NavLink } from 'react-router-dom';
import './card.css';

interface ICard {
  data: IPeople;
}

export const Card: FC<ICard> = ({ data }) => {
  const { name, url } = data;
  const id = url.replace(BASE_URL, '');
  return (
    <NavLink
      to={`info/${id}`}
      className={({ isActive, isPending }) =>
        isActive
          ? 'card card__item active'
          : isPending
            ? 'card card__item pending'
            : 'card card__item'
      }
    >
      <div className="card__content">
        <h2 className="card__title">{name}</h2>
      </div>
      <div className="card__backdrop">{name}</div>
    </NavLink>
  );
};
