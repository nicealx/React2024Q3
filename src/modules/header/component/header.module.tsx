import React from 'react';
import './header.css';

export class Header extends React.Component {
  render() {
    return <header className="header"></header>;
  }
}
