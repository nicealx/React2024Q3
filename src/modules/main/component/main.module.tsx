import { FC, useContext } from 'react';
import { Cards, Loader, NotFound } from '../../../components';
import { AppContext } from '../../../store/AppContext';
import { Outlet } from 'react-router-dom';
import './main.css';

export const Main: FC = () => {
  const { isLoading, peoples, searchText } = useContext(AppContext);
  return (
    <main className="main">
      {isLoading ? (
        <Loader />
      ) : peoples.length ? (
        <>
          <div className="main__wrap">
            <Cards cards={peoples} />
          </div>
          <Outlet />
        </>
      ) : (
        <NotFound searchText={searchText} />
      )}
    </main>
  );
};
