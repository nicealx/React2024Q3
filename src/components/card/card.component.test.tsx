import { test, expect, describe } from 'vitest';
import { render, screen } from '@testing-library/react';
import { Card } from './card.component';
import { ICard } from '../../interfaces/interface';
import { MemoryRouter } from 'react-router-dom';

describe('Test card', () => {
  describe('Card data exist', () => {
    const fakeData: ICard = {
      data: {
        name: 'Name',
        url: '1',
      },
    };
    test('Card on display', () => {
      const { asFragment } = render(
        <MemoryRouter>
          <Card data={fakeData.data} />
        </MemoryRouter>,
      );

      expect(asFragment()).toMatchSnapshot();
      expect(screen.findByText(/Name/i)).toBeDefined();
    });
  });
});
