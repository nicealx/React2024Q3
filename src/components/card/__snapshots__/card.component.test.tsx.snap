// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test card > Card data exist > Card on display 1`] = `
<DocumentFragment>
  <a
    class="card card__item"
    href="/info/1"
  >
    <div
      class="loader"
    />
    <div
      class="card__content"
    >
      <h2
        class="card__title"
      >
        Name
      </h2>
    </div>
    <div
      class="card__backdrop"
    >
      Name
    </div>
  </a>
</DocumentFragment>
`;

exports[`Test card > Card data exist > Info card on display 1`] = `
<DocumentFragment>
  <main
    class="main"
  >
    <div
      class="loader"
    />
  </main>
</DocumentFragment>
`;

exports[`Test card > Card data exist > Info card on display 2`] = `
<DocumentFragment>
  <a
    class="card card__item"
    href="/info/1"
  >
    <div
      class="loader"
    />
    <div
      class="card__content"
    >
      <h2
        class="card__title"
      >
        Name
      </h2>
    </div>
    <div
      class="card__backdrop"
    >
      Name
    </div>
  </a>
</DocumentFragment>
`;
